File: Algorithm.CSharp/BasicTemplateEurexFuturesAlgorithm.cs
/*
 * QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
 * Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
*/

using System;
using System.Collections.Generic;
using QuantConnect.Data;
using QuantConnect.Data.UniverseSelection;
using QuantConnect.Interfaces;
using QuantConnect.Orders;
using QuantConnect.Securities;
using QuantConnect.Securities.Future;

namespace QuantConnect.Algorithm.CSharp
{
    /// <summary>
    /// This algorithm tests and demonstrates EUREX futures subscription and trading:
    /// - It tests contracts rollover by adding a continuous future and asserting that mapping happens at some point.
    /// - It tests basic trading by buying a contract and holding it until expiration.
    /// - It tests delisting and asserts the holdings are liquidated after that.
    /// </summary>
    public class BasicTemplateEurexFuturesAlgorithm : QCAlgorithm, IRegressionAlgorithmDefinition
    {
        private Future _continuousContract;
        private Symbol _mappedSymbol;
        private Symbol _contractToTrade;
        private int _mappingsCount;
        private decimal _boughtQuantity;
        private decimal _liquidatedQuantity;
        private bool _delisted;

        public override void Initialize()
        {
            SetStartDate(2024, 5, 30);
            SetEndDate(2024, 6, 23);

            SetAccountCurrency(Currencies.EUR);
            SetCash(1000000);

            _continuousContract = AddFuture(Futures.Indices.EuroStoxx50, Resolution.Minute,
                dataNormalizationMode: DataNormalizationMode.BackwardsRatio,
                dataMappingMode: DataMappingMode.FirstDayMonth,
                contractDepthOffset: 0);
            _continuousContract.SetFilter(TimeSpan.Zero, TimeSpan.FromDays(180));
            _mappedSymbol = _continuousContract.Mapped;

            var benchmark = AddIndex("SX5E");
            SetBenchmark(benchmark.Symbol);

            var seeder = new FuncSecuritySeeder(GetLastKnownPrices);
            SetSecurityInitializer(security => seeder.SeedSecurity(security));
        }

        public override void OnData(Slice slice)
        {
            foreach (var changedEvent in slice.SymbolChangedEvents.Values)
            {
                if (++_mappingsCount > 1)
                {
                    throw new RegressionTestException($"{Time} - Unexpected number of symbol changed events (mappings): {_mappingsCount}. " +
                        $"Expected only 1.");
                }

                Debug($"{Time} - SymbolChanged event: {changedEvent}");

                if (changedEvent.OldSymbol != _mappedSymbol.ID.ToString())
                {
                    throw new RegressionTestException($"{Time} - Unexpected symbol changed event old symbol: {changedEvent}");
                }

                if (changedEvent.NewSymbol != _continuousContract.Mapped.ID.ToString())
                {
                    throw new RegressionTestException($"{Time} - Unexpected symbol changed event new symbol: {changedEvent}");
                }

                // Let's trade the previous mapped contract, so we can hold it until expiration for testing
                // (will be sooner than the new mapped contract)
                _contractToTrade = _mappedSymbol;
                _mappedSymbol = _continuousContract.Mapped;
            }

            // Let's trade after the mapping is done
            if (_contractToTrade != null && _boughtQuantity == 0 && Securities[_contractToTrade].Exchange.ExchangeOpen)
            {
                Buy(_contractToTrade, 1);
            }

            if (_contractToTrade != null && slice.Delistings.TryGetValue(_contractToTrade, out var delisting))
            {
                if (delisting.Type == DelistingType.Delisted)
                {
                    _delisted = true;

                    if (Portfolio.Invested)
                    {
                        throw new RegressionTestException($"{Time} - Portfolio should not be invested after the traded contract is delisted.");
                    }
                }
            }
        }

        public override void OnOrderEvent(OrderEvent orderEvent)
        {
            if (orderEvent.Symbol != _contractToTrade)
            {
                throw new RegressionTestException($"{Time} - Unexpected order event symbol: {orderEvent.Symbol}. Expected {_contractToTrade}");
            }

            if (orderEvent.Direction == OrderDirection.Buy)
            {
                if (orderEvent.Status == OrderStatus.Filled)
                {
                    if (_boughtQuantity != 0 && _liquidatedQuantity != 0)
                    {
                        throw new RegressionTestException($"{Time} - Unexpected buy order event status: {orderEvent.Status}");
                    }
                    _boughtQuantity = orderEvent.Quantity;
                }
            }
            else if (orderEvent.Direction == OrderDirection.Sell)
            {
                if (orderEvent.Status == OrderStatus.Filled)
                {
                    if (_boughtQuantity <= 0 && _liquidatedQuantity != 0)
                    {
                        throw new RegressionTestException($"{Time} - Unexpected sell order event status: {orderEvent.Status}");
                    }
                    _liquidatedQuantity = orderEvent.Quantity;

                    if (_liquidatedQuantity != -_boughtQuantity)
                    {
                        throw new RegressionTestException($"{Time} - Unexpected liquidated quantity: {_liquidatedQuantity}. Expected: {-_boughtQuantity}");
                    }
                }
            }
        }

        public override void OnSecuritiesChanged(SecurityChanges changes)
        {
            foreach (var addedSecurity in changes.AddedSecurities)
            {
                if (addedSecurity.Symbol.SecurityType == SecurityType.Future && addedSecurity.Symbol.IsCanonical())
                {
                    _mappedSymbol = _continuousContract.Mapped;
                }
            }
        }

        public override void OnEndOfAlgorithm()
        {
            if (_mappingsCount == 0)
            {
                throw new RegressionTestException($"Unexpected number of symbol changed events (mappings): {_mappingsCount}. Expected 1.");
            }

            if (!_delisted)
            {
                throw new RegressionTestException("Contract was not delisted");
            }

            // Make sure we traded and that the position was liquidated on delisting
            if (_boughtQuantity <= 0 || _liquidatedQuantity >= 0)
            {
                throw new RegressionTestException($"Unexpected sold quantity: {_boughtQuantity} and liquidated quantity: {_liquidatedQuantity}");
            }
        }

        /// <summary>
        /// This is used by the regression test system to indicate if the open source Lean repository has the required data to run this algorithm.
        /// </summary>
        public bool CanRunLocally { get; } = true;

        /// <summary>
        /// This is used by the regression test system to indicate which languages this algorithm is written in.
        /// </summary>
        public List<Language> Languages { get; } = new() { Language.CSharp, Language.Python };

        /// <summary>
        /// Data Points count of all timeslices of algorithm
        /// </summary>
        public long DataPoints => 94326;

        /// <summary>
        /// Data Points count of the algorithm history
        /// </summary>
        public int AlgorithmHistoryDataPoints => 0;

        /// <summary>
        /// Final status of the algorithm
        /// </summary>
        public AlgorithmStatus AlgorithmStatus => AlgorithmStatus.Completed;

        /// <summary>
        /// This is used by the regression test system to indicate what the expected statistics are from running the algorithm
        /// </summary>
        public Dictionary<string, string> ExpectedStatistics => new Dictionary<string, string>
        {
            {"Total Orders", "2"},
            {"Average Win", "0%"},
            {"Average Loss", "-0.11%"},
            {"Compounding Annual Return", "-1.667%"},
            {"Drawdown", "0.100%"},
            {"Expectancy", "-1"},
            {"Start Equity", "1000000"},
            {"End Equity", "998849.48"},
            {"Net Profit", "-0.115%"},
            {"Sharpe Ratio", "-34.455"},
            {"Sortino Ratio", "-57.336"},
            {"Probabilistic Sharpe Ratio", "0.002%"},
            {"Loss Rate", "100%"},
            {"Win Rate", "0%"},
            {"Profit-Loss Ratio", "0"},
            {"Alpha", "0"},
            {"Beta", "0"},
            {"Annual Standard Deviation", "0.002"},
            {"Annual Variance", "0"},
            {"Information Ratio", "-6.176"},
            {"Tracking Error", "0.002"},
            {"Treynor Ratio", "0"},
            {"Total Fees", "€1.02"},
            {"Estimated Strategy Capacity", "€2300000000.00"},
            {"Lowest Capacity Asset", "FESX YJHOAMPYKRS5"},
            {"Portfolio Turnover", "0.40%"},
            {"OrderListHash", "ac9acc478ba1afe53993cdbb92f8ec6e"}
        };
    }
}

File: Algorithm.Python/BasicTemplateEurexFuturesAlgorithm.py
# QUANTCONNECT.COM - Democratizing Finance, Empowering Individuals.
# Lean Algorithmic Trading Engine v2.0. Copyright 2014 QuantConnect Corporation.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from AlgorithmImports import *

### <summary>
### This algorithm tests and demonstrates EUREX futures subscription and trading:
### - It tests contracts rollover by adding a continuous future and asserting that mapping happens at some point.
### - It tests basic trading by buying a contract and holding it until expiration.
### - It tests delisting and asserts the holdings are liquidated after that.
### </summary>
class BasicTemplateEurexFuturesAlgorithm(QCAlgorithm):
    def __init__(self):
        super().__init__()
        self._continuous_contract = None
        self._mapped_symbol = None
        self._contract_to_trade = None
        self._mappings_count = 0
        self._bought_quantity = 0
        self._liquidated_quantity = 0
        self._delisted = False

    def initialize(self):
        self.set_start_date(2024, 5, 30)
        self.set_end_date(2024, 6, 23)

        self.set_account_currency(Currencies.EUR);
        self.set_cash(1000000)

        self._continuous_contract = self.add_future(
            Futures.Indices.EURO_STOXX_50,
            Resolution.MINUTE,
            data_normalization_mode=DataNormalizationMode.BACKWARDS_RATIO,
            data_mapping_mode=DataMappingMode.FIRST_DAY_MONTH,
            contract_depth_offset=0,
        )
        self._continuous_contract.set_filter(timedelta(days=0), timedelta(days=180))
        self._mapped_symbol = self._continuous_contract.mapped

        benchmark = self.add_index("SX5E")
        self.set_benchmark(benchmark.symbol)

        func_seeder = FuncSecuritySeeder(self.get_last_known_prices)
        self.set_security_initializer(lambda security: func_seeder.seed_security(security))

    def on_data(self, slice):
        for changed_event in slice.symbol_changed_events.values():
            self._mappings_count += 1
            if self._mappings_count > 1:
                raise AssertionError(f"{self.time} - Unexpected number of symbol changed events (mappings): {self._mappings_count}. Expected only 1.")

            self.debug(f"{self.time} - SymbolChanged event: {changed_event}")

            if changed_event.old_symbol != str(self._mapped_symbol.id):
                raise AssertionError(f"{self.time} - Unexpected symbol changed event old symbol: {changed_event}")

            if changed_event.new_symbol != str(self._continuous_contract.mapped.id):
                raise AssertionError(f"{self.time} - Unexpected symbol changed event new symbol: {changed_event}")

            # Let's trade the previous mapped contract, so we can hold it until expiration for testing
            # (will be sooner than the new mapped contract)
            self._contract_to_trade = self._mapped_symbol
            self._mapped_symbol = self._continuous_contract.mapped

        # Let's trade after the mapping is done
        if self._contract_to_trade is not None and self._bought_quantity == 0 and self.securities[self._contract_to_trade].exchange.exchange_open:
            self.buy(self._contract_to_trade, 1)

        if self._contract_to_trade is not None and slice.delistings.contains_key(self._contract_to_trade):
            delisting = slice.delistings[self._contract_to_trade]
            if delisting.type == DelistingType.DELISTED:
                self._delisted = True

                if self.portfolio.invested:
                    raise AssertionError(f"{self.time} - Portfolio should not be invested after the traded contract is delisted.")

    def on_order_event(self, order_event):
        if order_event.symbol != self._contract_to_trade:
            raise AssertionError(f"{self.time} - Unexpected order event symbol: {order_event.symbol}. Expected {self._contract_to_trade}")

        if order_event.direction == OrderDirection.BUY:
            if order_event.status == OrderStatus.FILLED:
                if self._bought_quantity != 0 and self._liquidated_quantity != 0:
                    raise AssertionError(f"{self.time} - Unexpected buy order event status: {order_event.status}")

                self._bought_quantity = order_event.quantity
        elif order_event.direction == OrderDirection.SELL:
            if order_event.status == OrderStatus.FILLED:
                if self._bought_quantity <= 0 and self._liquidated_quantity != 0:
                    raise AssertionError(f"{self.time} - Unexpected sell order event status: {order_event.status}")

                self._liquidated_quantity = order_event.quantity
                if self._liquidated_quantity != -self._bought_quantity:
                    raise AssertionError(f"{self.time} - Unexpected liquidated quantity: {self._liquidated_quantity}. Expected: {-self._bought_quantity}")

    def on_securities_changed(self, changes):
        for added_security in changes.added_securities:
            if added_security.symbol.security_type == SecurityType.FUTURE and added_security.symbol.is_canonical():
                self._mapped_symbol = self._continuous_contract.mapped

    def on_end_of_algorithm(self):
        if self._mappings_count == 0:
            raise AssertionError(f"Unexpected number of symbol changed events (mappings): {self._mappings_count}. Expected 1.")

        if not self._delisted:
            raise AssertionError("Contract was not delisted")

        # Make sure we traded and that the position was liquidated on delisting
        if self._bought_quantity <= 0 or self._liquidated_quantity >= 0:
            raise AssertionError(f"Unexpected sold quantity: {self._bought_quantity} and liquidated quantity: {self._liquidated_quantity}")
